components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/JDBCPermissionBasedInternalScopeValidator.java

```diff
@@ -180,11 +184,24 @@ Note that if there is association between a federated user and local user () 'As
              */
             if (authenticatedUser.isFederatedUser()) {
                 /*
-                There is a flow where 'Assert identity using mapped local subject identifier' flag enabled but the
-                federated user doesn't have any association in localIDP, to handle this case we check for 'Assert
-                identity using mapped local subject identifier' flag and get roles from userStore.
+                 * If the role based authorization feature is enabled & particular applications in the listed under the
+                 * required application list then retrieve permissions from the FIdp user roles.
+                 * Since tenant flow is enabled, permission will be fetched from user login(federated) tenant
+                 * not the application tenant.
                  */
-                if (isSPAlwaysSendMappedLocalSubjectId(clientId)) {
+                if (Boolean.parseBoolean(IdentityUtil.getProperty(
+                        IdentityConstants.SystemRoles.ENABLE_FEDERATED_IDP_ROLE_BASED_AUTHORIZATION))
+                        && IdentityUtil.getFederatedRoleBasedAuthorizationEnabledApplications().contains((
+                        OAuth2Util.getServiceProvider(clientId).getApplicationName()))) {
+                    allowedUIResourcesForUser =
+                            getAllowedUIResourcesForNonAssociatedFederatedUserWithRoles(
```

Comment:
`getAllowedUIResourcesForNonAssociatedFederatedUserWithRoles` better to rename this